app/src/main/java/ch/hikemate/app/model/elevation/ElevationRepositoryCopernicus.kt

```diff
@@ -0,0 +1,399 @@
+   *
+   * @param onSuccess The callback to be called when the request is successful
+   * @param onFailure The callback to be called when the request fails
+   */
+  private fun launchUpdateOfCache(
+      onSuccess: (List<Double>) -> Unit = {},
+      onFailure: (Exception) -> Unit = {}
+  ) {
+
+    if (requests.isEmpty()) return
+
+    // We need to start a new job to update the cache
+    cacheUpdateJob =
+        CoroutineScope(repoDispatcher).launch {
+          // Delay the update by some amount of time to allow for multiple requests
+          // to be batched together
+          // This is purely empirical and can be adjusted
+          // If we already have enough data for a request, we don't need to wait
+          var coordinates: Set<LatLong> = emptySet()
+          val currentlyProcessedRequests = mutableSetOf<ElevationRequest>()
+          mutex.withLock {
+            coordinates = requests.flatMap { it.coordinates }.toSet()
+            currentlyProcessedRequests.addAll(requests)
+          }
+          if (coordinates.size < MAX_COORDINATES_PER_REQUEST) {
+            delay(WAITING_DELAY_BEFORE_SENDING_REQUEST_FOR_BATCHING)
+            // If there were new requests in the meantime, we need to update the coordinates
+            mutex.withLock {
+              coordinates = requests.flatMap { it.coordinates }.toSet()
+              currentlyProcessedRequests.clear()
+              currentlyProcessedRequests.addAll(requests)
+            }
+          }
+          mutex.withLock {
+            val chunks = coordinates.chunked(MAX_COORDINATES_PER_REQUEST)
+
+            for (chunk in chunks) {
+              // Create the JSON body for the request
+              val jsonString = buildJsonArray {
+                chunk.forEach { coordinate ->
+                  add(
+                      buildJsonArray {
+                        add(coordinate.lat)
+                        add(coordinate.lon)
+                      })
+                }
+              }
+
+              val request =
+                  okhttp3.Request.Builder().url(BASE_URL + jsonString.toString()).get().build()
+
+              // Callback for the network request, it also updates the cache
+              val onSuccessWithCache: (List<Double?>) -> Unit =
+                  parseCacheOnDataRetrieval(chunk, onSuccess)
+
+              val onFailureCallback: (Exception) -> Unit = { exception ->
+                CoroutineScope(repoDispatcher).launch {
+                  mutex.withLock {
+                    currentlyProcessedRequests
+                        .filter { it.coordinates.containsAll(chunk) }
+                        .forEach { request ->
+                          requests.remove(request)
+                          request.onFailure(exception)
+                        }
+                  }
+                  onFailure(exception)
+                }
+              }
```

Comment:
Suggestion: beautiful, you fixed the case where new requests would be marked as failed with the others 👍🏻

One thing though, it makes more sense imo to call `onFailure` without having the lock, because the control is given back to the caller, thus the UI, and things might be done that will keep the lock for a while unnecessarily.

Reply:
Actually, this specific piece of code has been removed during my 2nd refactor. I've added Coroutines around other `onSuccess` in mutex!